### ISSUE TITLE
Include namespace in ui/metrics

### ISSUE THREAD

**Issue by runningman84** [opencost/opencost#1964]
**Is your feature request related to a problem? Please describe.**

I am using the opencost ui and I do a breakdown based on this:
controller,service,pod
Unfortunately  the listing does not contain the namespace.

**Describe the solution you'd like**

For example instead of the current listing like:
deployment:traefik
I would rather like to see
deployment:traefik:trafik

because different teams could deploy the same app in their own namepsace like
deployment:team-a:traefik
deployment:team-b:traefik

**Describe alternatives you've considered**
I have not found any option to include the namespace.

**Additional context**
Add any other context, documentation links, or screenshots about the feature request here.

**Comment by mattray** [comment]
It sounds like you're looking for multi-aggregation? It's supported in the API but not the UI (yet)
https://github.com/opencost/opencost-ui/issues/9

**Comment by runningman84** [comment]
yes that works fine... one additional nice feature would be to habe a filter similar to the aggregation. Eg. only show pods in a given namespace...

**Comment by jwcesign** [comment]
Can I take this? @mattray

**Comment by joao-pimentel-randoli** [comment]
Just to provide some more details, what the OP requested can be achieved by using multi-aggregation like this: `agg=namespace,controller`

As Matt had mentioned, this can't be done directly through the UI, you need to manually edit the URL. Here's an example in the demo environment: https://demo.infra.opencost.io/?agg=namespace,controller

<img width="1536" height="1004" alt="Image" src="https://github.com/user-attachments/assets/9a1d7288-5e35-4aa9-9fb4-cb244122483d" />

**Comment by joao-pimentel-randoli** [comment]
@ameijer pls consider closing this and reopening the UI issue: https://github.com/opencost/opencost-ui/issues/9

**Comment by Kush172005** [comment]
Hi @mattray , I’d like to work on this issue. Since the backend already supports `agg=namespace,controller`, I’m planning to update the UI so it can include namespace in the controller breakdown. Let me know if that approach sounds good.

**Comment by vincent067** [comment]
Hey there! 👋

First off, thanks for building OpenCost - it's an awesome tool for Kubernetes cost monitoring! 

I've been looking into this issue and I think I understand the pain point. When multiple teams deploy the same app (like Traefik) in different namespaces, the UI currently shows:

```
deployment:traefik
deployment:traefik
```

Which makes it hard to tell which is which 😅

After digging through the codebase, I found that the aggregation key is generated in `core/pkg/opencost/allocationprops.go` in the `GenerateKey()` function (around line 308). Currently for controller aggregation, it formats as:

```go
controller = fmt.Sprintf("%s:%s", p.ControllerKind, controller)
```

So it produces `deployment:traefik` but doesn't include the namespace.

**My question:** Would the maintainers be open to a PR that adds namespace to the controller key format? Something like:

```go
controller = fmt.Sprintf("%s:%s:%s", p.ControllerKind, p.Namespace, controller)
```

This would produce `deployment:traefik:traefik` (or with your example: `deployment:team-a:traefik` and `deployment:team-b:traefik`).

I realize this would be a breaking change for anyone relying on the current format, so maybe it could be:
1. A configurable option via the OpenCost config
2. A new aggregation property like `controllerWithNamespace`
3. Or simply documented as a breaking change in the next release

What do you think? I'm happy to work on a PR if there's agreement on the approach! 🙏

Also, the same consideration might apply to `service` and `pod` aggregations mentioned in the issue title - I can look into those too if helpful.

Thanks for your time! 😊

**Comment by vincent067** [comment]
Hey there! 👋 

I was looking at this issue and wanted to share some findings that might help move it forward.

## The Problem

When aggregating by acontrollera, the current implementation generates keys in the format adeployment:traefika without including the namespace. This makes it impossible to distinguish between deployments with the same name in different namespaces (e.g., ateam-a/traefika vs ateam-b/traefika).

## Where the Change Would Go

The key generation happens in jcore/pkg/opencost/allocationprops.goa in the jGenerateKey()a function around line 405-413:

jjgo
case agg == AllocationControllerProp:
    controller := p.Controller
    if controller ==  {
        controller = UnallocatedSuffix
    } else if p.ControllerKind !=  {
        controller = fmt.Sprintf(%s:%s, p.ControllerKind, controller)
    }
    names = append(names, controller)
jjj

## Potential Solutions

### Option 1: Include namespace in controller key (Breaking Change)
Change the format to jdeployment:namespace:namea (e.g., jdeployment:team-a:traefika).

### Option 2: Add a new aggregation property (Non-breaking)
Add a new property like jcontrollerWithNamespacea that includes the namespace, keeping the existing jcontrollera property unchanged.

### Option 3: Multi-level aggregation (Recommended)
If you aggregate by both jnamespacea and jcontrollera, you'd get the grouping you need:
- First level: namespace (team-a, team-b)
- Second level: controller (deployment:traefik)

## Quick Workaround

For now, if you're using the API directly, you can aggregate by jnamespace,controllera instead of just jcontrollera. This will give you a hierarchical breakdown that distinguishes between namespaces.

I'd be happy to submit a PR for Option 2 if the maintainers think it's a good approach. It would add flexibility without breaking existing integrations. What do you all think? 🤔

cc @opencost/maintainers

**Comment by vincent067** [comment]
Hi @runningman84! 👋

Great feature request! I totally understand the pain point - when multiple teams deploy the same app (like traefik) in different namespaces, the current breakdown becomes really confusing. 😅

**Root Cause Analysis:**
I took a quick look at the codebase and found that the aggregation key is generated in `core/pkg/opencost/allocationprops.go` in the `GenerateKey()` function. Currently, when aggregating by controller, it formats the key as:

```go
controller = fmt.Sprintf("%s:%s", p.ControllerKind, controller)
```

This produces `deployment:traefik` without the namespace context.

**Possible Solutions:**

1. **Modify `GenerateKey()` to include namespace** when `AllocationControllerProp` is in the aggregation list. Something like:
   ```go
   controller = fmt.Sprintf("%s:%s:%s", p.ControllerKind, p.Namespace, controller)
   ```
   This would produce: `deployment:traefik:team-a` (following your suggested format)

2. **Add a new aggregation option** like `controllerWithNamespace` to maintain backward compatibility for users who prefer the current behavior.

**Considerations:**
- The `controllerKey` struct already includes namespace (see `pkg/costmodel/key.go`), so the data is available
- This change might affect existing users who rely on the current naming format in their reports/dashboards
- The UI might need minor adjustments to handle longer display names

I'd be happy to help test any implementation! This would be a really useful addition for multi-tenant clusters. 🚀

cc @opencost-maintainers - any thoughts on the preferred approach? Would you prefer a breaking change to the default format or a new aggregation option?

**Comment by vincent067** [comment]
Hey! Was poking around the codebase while looking at this issue and thought I'd share what I found.

Looking at `core/pkg/opencost/allocationprops.go`, the `GenerateKey` function handles how aggregation keys are built. When you aggregate by `controller`, it currently formats it as `controllerKind:controller` (e.g., `deployment:traefik`). The namespace is available in the properties but isn't included in that specific key format.

I see a couple of possible approaches:

1. **Modify the controller aggregation format** - Update `GenerateKey` to include namespace when generating controller keys, something like `deployment:namespace:traefik`. Though this might break existing integrations that depend on the current format.

2. **Multi-aggregation** - As @mattray mentioned, using multiple aggregation fields like `namespace,controller,service,pod` would give you the breakdown structure. The UI might just need better handling for displaying multi-level aggregations.

The API already supports multi-aggregation via the `aggregate` query param (comma-separated), so this seems like mainly a UI presentation issue?

I'm still getting familiar with the UI side - is there a specific repo for the UI components? I saw the mention of opencost-ui but wasn't sure if that's where the breakdown display logic lives.

Happy to dig deeper if someone can point me in the right direction!

- Ming

**Comment by vincent067** [comment]
Hey there! 👋

I was looking into this issue and I think I understand the pain point. When you aggregate by `controller`, you get entries like `deployment:traefik` but there's no way to tell which namespace that deployment belongs to. This is definitely confusing when multiple teams deploy the same app in different namespaces!

I was poking around the codebase and found that in `core/pkg/opencost/allocationprops.go`, the `GenerateKey` method handles the controller aggregation:

```go
case agg == AllocationControllerProp:
    controller := p.Controller
    if controller == "" {
        controller = UnallocatedSuffix
    } else if p.ControllerKind != "" {
        controller = fmt.Sprintf("%s:%s", p.ControllerKind, controller)
    }
    names = append(names, controller)
```

So currently it formats as `controllerKind:controllerName`, but the namespace isn't included.

A few ideas:

1. **Include namespace in the key**: Format as `controllerKind:namespace:controllerName` (e.g., `deployment:team-a:traefik`). This would be a breaking change though.

2. **Add a new aggregation option**: Something like `controllerWithNamespace` that includes the namespace.

3. **UI enhancement**: Display namespace as a separate column when hovering/expanding.

I'd be happy to help implement option 1 or 2 if that sounds reasonable! Also, as a workaround, you could try aggregating by `namespace,controller` which gives separate rows per namespace.

Cheers! 😊
- @vincent067
